### PR TITLE
TimeSeriesProperty removing filterByTime and filterByTimes

### PR DESCRIPTION
**Pull request by peterfpeterson**
Do not merge before #35440.

This removes the dependency of `GetAllEi` and `SumEventsByLogValue` on `TimeSeriesProperty::filterByTime()` and `TimeSeriesProperty::filterByTimes()` then removes `TimeSeriesProperty::filterByTimes()` which was previously deprecated. There is an additional test in `TimeSeriesProperty` to show off an issue with log times that are often used in test examples.

**To test:**

Check out the new unit test and look over the changes in `GetAllEi`.

*There is no associated issue*, but this is part of #34794.

*This does not require release notes* because it is already covered in previous release notes... and it is a refactor.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/CalculateCountRate.cpp [hunk 1]
```diff
@@ -384,9 +385,18 @@ void CalculateCountRate::setOutLogParameters(const DataObjects::EventWorkspace_s
         if (!m_tmpLogHolder) {
           m_tmpLogHolder = std::unique_ptr<Kernel::TimeSeriesProperty<double>>(m_pNormalizationLog->clone());
         }
-        m_tmpLogHolder->filterByTime(runTMin, runTMax);
-        m_pNormalizationLog = m_tmpLogHolder.get();
-        m_numLogSteps = m_pNormalizationLog->realSize();
+        Kernel::TimeROI roi(runTMin, runTMax);
+        if (Kernel::FilteredTimeSeriesProperty<double> *filteredLog =
+                dynamic_cast<Kernel::FilteredTimeSeriesProperty<double> *>(m_tmpLogHolder.get())) {
+          filteredLog->filterWith(roi);
```
Comment (jmborr) [inline Framework/Algorithms/src/CalculateCountRate.cpp:391]: (rant) this `filterWith` name is unfortunate. I thought the function was actually filtering the time series, but actually is just replacing its internal filter.

File: Framework/Algorithms/src/GetAllEi.cpp [hunk 2]
```diff
@@ -914,19 +914,26 @@ double GetAllEi::getAvrgLogValue(const API::MatrixWorkspace_sptr &inputWS, const
...
-    auto TimeStart = inputWS->run().startTime();
-    auto TimeEnd = inputWS->run().endTime();
-    pTimeSeries->filterByTime(TimeStart, TimeEnd);
+    const auto timeStart = inputWS->run().startTime();
+    const auto timeStop = inputWS->run().endTime();
+    if (timeStart < timeStop) {
+      Kernel::TimeROI localROI(timeStart, timeStop);
+      value = pTimeSeries->getStatistics(&localROI).mean;
+    } else {
+      value = pTimeSeries->getStatistics().mean;
```
Comment (jmborr) [inline Framework/Algorithms/src/GetAllEi.cpp:925]:
```suggestion
       // all values in pTimeSeries will be used
      value = pTimeSeries->getStatistics().mean;
```

**Review by jmborr — CHANGES_REQUESTED** [review]
Lot of good work, congratulations :tada: 
I only have minor comments

File: Framework/Algorithms/src/CalculateCountRate.cpp (hunk 1 above)
Comment (peterfpeterson) [reply Framework/Algorithms/src/CalculateCountRate.cpp:391]: The name is even in the tutorials. We're stuck on this one.

File: Framework/Kernel/src/TimeROI.cpp [hunk 3]
```diff
@@ -487,6 +494,27 @@ const std::vector<Kernel::TimeInterval> TimeROI::toTimeIntervals() const {
   return output;
 }
 
+/**
+ * Time intervals returned where no time is before after. This is used in calculating ranges
+ * in TimeSeriesProperty.
+ */
+const std::vector<Kernel::TimeInterval> TimeROI::toTimeIntervals(const Types::Core::DateAndTime &after) const {
```
[the remark this answers is not in the document]
Comment (peterfpeterson) [reply Framework/Kernel/src/TimeROI.cpp:501]: That would clash with `toTimeIntervals()` which already exists.

**Review by jmborr — APPROVED** [review]

**Pull request merged**